stage0/gc.go

```diff
@@ -65,55 +66,94 @@ type mount struct {
 	id         int
 	parentId   int
 	mountPoint string
+	mode       string
 }
 
 type mounts []*mount
 
-// Less ensures that mounts are sorted in an order we can unmount; child before
-// parent. This works by pushing all child mounts up to the front. Parent
-// mounts are halted once they encounter a child. The actual sequential order
-// of the mounts doesn't matter; only that children are in front of parents.
-func (m mounts) Less(i, j int) bool { return m[i].id != m[j].parentId }
-func (m mounts) Len() int           { return len(m) }
-func (m mounts) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+// getMountDepth searches the list for all ancestors of the mount at index i
+func (m mounts) getMountDepth(i int) int {
+	ancestorMap := make(map[int]struct{}, 0)
+	current := m[i]
+	for found := true; found; {
+		found = false
+		for _, mnt := range m {
+			if mnt.id == current.parentId {
+				ancestorMap[mnt.id] = struct{}{}
+				current = mnt
+				found = true
+				break
+			}
+		}
+	}
+	return len(ancestorMap)
+}
+
+// Less ensures that mounts are sorted in an order we can unmount; descendant before ancestor.
+// The requirement of transitivity for Less has to be fulfilled otherwise the sort algorithm will fail.
+func (m mounts) Less(i, j int) (result bool) { return m.getMountDepth(i) >= m.getMountDepth(j) }
+func (m mounts) Len() int                    { return len(m) }
+func (m mounts) Swap(i, j int)               { m[i], m[j] = m[j], m[i] }
 
 // getMountsForPrefix parses mi (/proc/PID/mountinfo) and returns mounts for path prefix
 func getMountsForPrefix(path string, mi io.Reader) (mounts, error) {
 	var podMounts mounts
 	sc := bufio.NewScanner(mi)
+	var (
+		mountId    int
+		parentId   int
+		mountPoint string
+		mode       string
+	)
+
+	//	lineRegex := regexp.MustCompile(`[^ ]+`)
 	for sc.Scan() {
-		var (
-			mountId    int
-			parentId   int
-			discard    string
-			mountPoint string
-		)
-
-		_, err := fmt.Sscanf(sc.Text(), "%d %d %s %s %s",
-			&mountId, &parentId, &discard, &discard, &mountPoint)
-		if err != nil {
-			return nil, err
+		line := sc.Text()
+		lineResult := strings.Split(line, " ")
+		if len(lineResult) < 7 {
+			return nil, fmt.Errorf("Not enough fields from line %q: %+v", line, lineResult)
+		}
+
+		for i, s := range lineResult {
+			if s == "-" {
+				break
+			}
+			var err error
+			switch i {
+			case 0:
+				mountId, err = strconv.Atoi(s)
+			case 1:
+				parentId, err = strconv.Atoi(s)
+			case 2, 3:
+			case 4:
+				mountPoint = s
+			default:
+				mode += s + ","
```

Comment:
I think this comma separated field and then `strings.Contains` is not ideal, but it's ok in this PR. It could be revisited later if we actually need more precise parsing.